Allows to use realtime subscriptions in plugins (#1645)
kuzzleio/kuzzle: doc/2/plugins/plugin-context/accessors/sdk/index.md

Realtime subscription should be made in the plugin [init](core/2/plugins/guides/manual-setup/init-function) method or in a hook on the [kuzzle:start:before](/core/2/plugins/guides/events/kuzzle-start) event.

::: warning
You should avoid making subscription at runtime because that can lead to unwanted behavior since the subscriptions will not be replicated on every cluster node.
:::

The `propagate` option defines if, for that subscription, notifications should be propagated to (and processed by) all cluster nodes, or if only the node having received the triggering event should handle it.

By default When you receive a notification, only one node will execute the associated callback.  

#### propagate: false (default)

With `propagate: false`, the callback function will be executed only on the node receiving a request who triggers a notification. (only one execution) 

::: info 
This behavior is suitable for most usage like sending emails, write in the database, call an external API, etc.
:::

#### propagate: true

> With `propagate: true`, the callback function will be executed one each node of the cluster. (n executions)

Reviewer: ```suggestion
With `propagate: true`, notifications are propagated to all nodes of a cluster, executing all callback functions.
```

Author: That's not true, the notifications are always propagated but they will not trigger notifications but since it's an internal behavior I'm OK with this sentence if it's more easy to understand for the users

::: info 
This behavior is suitable for synchronizing RAM cache amongst cluster nodes for example.
:::

#### Example

```js
async init (config, context) {
  context.accessors.sdk.realtime.subscribe(
    'nyc-open-data',
    'yellow-taxi',
    {},
    notification => {

>       // this callback will be executed only on one node

Reviewer: ```suggestion
      // this callback will be executed only on the node generating a notification
```

    });

>   // the default value for the "propagate" option is "true"

Reviewer: Isn't it `false` by default?

- The created document will have the `author` metadata property set to `null`.

## Realtime notifications

<SinceBadge version="change-me" />

The [realtime](/sdk/js/7/controllers/realtime) is entirely available with the [subscribe](/sdk/js/7/controllers/realtime/subscribe) and [unsubscribe](/sdk/js/7/controllers/realtime/unsubscribe) methods.  

Realtime subscription should be made in the plugin [init](core/2/plugins/guides/manual-setup/init-function) method or in a hook on the [kuzzle:start:before](/core/2/plugins/guides/events/kuzzle-start) event.

::: warning

> If you want to make subscription in hooks, pipes, or controller actions then you should use the `replicate: true` option otherwise your callback will not be executed with every notification.

Reviewer: Hard to understand, and possibly a source of many support tickets.

I have multiple possible suggestions to improve the wording of this warning, but first I need to understand something that bothers me:
* when `replicate: true` is set, what is replicated? The subscription, or notifications resulting from it?
* in the latter case, will a request submitted on a different node:
  * trigger the notification on the node where the plugin made the subscription?
  * trigger the notification on the same node than where the notification is generated (the subscription is replicated but the notification is not propagated)?
  * don't trigger the notification

Author: Do you still need insight on this one since we discuss the matter ?

Reviewer: Nope I'm fine now. Thanks.

:::

When you receive a notification, only one node will execute the associated callback.  

::: info

>   You can use the `replicate` option if you want your callback function to be executed on each nodes.

Reviewer: I don't get it. If plugins subscribe before kuzzle starts, then the probability that they will register the same filters is close to 1. And that means that all plugin instances should receive the notification, independantly to the value of the `replicate` option.

Reviewer: I wonder if this is an architecture problem, or a semantic one: could `propagate` be a better keyword for this behavior?

Author: Proposal: `executionStrategy: cluster|node`

Reviewer: Also hard to understand. I might want a "node" behavior in a cluster, and this new proposal can lead to the assumption that this should not be used that way.

If I think about how to document the feature to users, I think that `propagate` is simpler to understand. We could write something like this:

```
The `propagate` option allows to define if, for that subscription, notifications should be propagated to all cluster nodes, or if only the node having received the triggering event should execute it.
Values: true, false
```

wdyt?

Author: LGTML :+1:

:::

#### replicate: false (default)

> With `replicate: false`, the callback function will be executed only on the node receiving a request who triggers a notification. (only one execution) 

Reviewer: Impractical: subscriptions should always be kept in sync in all nodes. The question is if a notification is propagated/duplicated or not. Am I right?

::: info 
This behavior is suitable for most usage like sending emails, write in the database, call an external API, etc.
:::

#### replicate: true

With `replicate: true`, the callback function will be executed one each node of the cluster. (n executions)

::: info 
This behavior is suitable for synchronizing RAM cache amongst cluster nodes for example.
:::

#### Example

```js
async init (config, context) {
  context.accessors.sdk.realtime.subscribe(
    'nyc-open-data',
    'yellow-taxi',
    {},
    notification => {
      // this callback will be executed only on one node
    });

>   // the default value for the "cluster" option is "true"

Reviewer 2: ```suggestion
  // the default value for the "replicate" option is "true"
```

Author: Good catch ;)
